ddtrace/internal/_encoding.pyx

```diff
@@ -46,120 +47,327 @@ cdef inline int PyBytesLike_Check(object o):
+        return self.pk.length - 5 + array_prefix_size(len(self._index))
+
+    cpdef append_raw(self, object src, Py_ssize_t size):
+        cdef int res
+        with self._lock:
+            assert self.size + size <= self.max_size
+            res = msgpack_pack_raw_body(&self.pk, <char *>PyLong_AsLong(src), size)
+            if res != 0:
+                raise RuntimeError("Failed to append raw bytes to msgpack string table")
+
+
+    cpdef flush(self):
+        try:
+            return self.get_bytes()
+        finally:
+            self.pk.length = 6
+            self._sp = 0
 
-        if PY_MAJOR_VERSION < 3:
-            self.encoding = "utf-8"
-        else:
-            self.encoding = NULL
+
+cdef class BufferedEncoder(object):
+    content_type: str = None
+    def __init__(self, max_size, max_item_size): ...
+    def __len__(self): ...
+    def put(self, trace): ...
+    def encode(self): ...
+
+
+cdef class MsgpackEncoder(BufferedEncoder):
+    content_type = "application/msgpack"
+
+    cdef public int max_size
+    cdef public int max_item_size
+    cdef msgpack_packer pk
+    cdef object _lock
+    cdef int _count
+
+    def __cinit__(self, max_size, max_item_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
+        if self.pk.buf == NULL:
+            raise MemoryError("Unable to allocate internal buffer.")
+
+        self.pk.buf_size = self.max_size = max_size
+        self.max_item_size = max_item_size if max_item_size < max_size else max_size
+        self._lock = threading.Lock()
+        self._count = 0
+
+        self.pk.length = 5  # Leave room for array length prefix
 
     def __dealloc__(self):
         PyMem_Free(self.pk.buf)
         self.pk.buf = NULL
 
-    cdef inline object _flush_buffer(self):
-        buf = PyBytes_FromStringAndSize(self.pk.buf, self.pk.length)
-        # Reset the buffer.
-        self.pk.length = 0
-        return buf
-
-    cdef inline int _pack_number(self, object n):
-        if n is None:
-            return msgpack_pack_nil(&self.pk)
-        if PyLong_Check(n):
-            # PyInt_Check(long) is True for Python 3.
-            # So we should test long before int.
-            try:
-                if n > 0:
-                    return msgpack_pack_unsigned_long_long(&self.pk, <unsigned long long> n)
-                return msgpack_pack_long_long(&self.pk, <long long> n)
-            except OverflowError as oe:
-                if n is not self._default:
-                    return self._pack_number(self._default)
-                raise OverflowError("Integer value out of range")
-
-        elif PyInt_Check(n):
-            return msgpack_pack_long(&self.pk, <long> n)
-
-        elif PyFloat_Check(n):
-            return msgpack_pack_double(&self.pk, <double> n)
-
-        raise TypeError("Unhandled numeric type: %r" % type(n))
-
-    cdef inline int _pack_text(self, object text):
+    def __len__(self):  # TODO: Do we need this?
+        return self._count
```

Comment:
Don't think we need this, but just my 2 cents 😄

Reply:
Just had a look around and saw that we have `len(self._encoder),` in `writer.py` to get the count of the traces currently held by the encoder. So perhaps we need this, but not necessarily in the form of a `__len__`. I'm not happy with this because the encoders are now buffered and `__len__` could easily be mistaken for `size` which gives you the number of bytes in the buffer. So maybe we still need this but with a better name 🙂